### ISSUE TITLE
bool& get_use_assertions() should not be thread safe

### ISSUE THREAD

**Issue by lrineau** [CGAL/cgal#7222]
## Issue Details

See the original discussion here: https://github.com/CGAL/cgal/commit/9b357c6afff4cb7e6175e9096ea49e8585dc9681#r97801479

In my opinion, `get_use_assertions()` should not be thread-safe, and the flag should be global: any user code wanting to toggle the use of assertions should ensure that no other thread is currently using CGAL code. That makes no sense to try toggle that flag during the execution of other threads.

**Comment by GilesBathgate** [comment]
I guess you could be nice to end users by adding std::lock_guard to the get/set methods. However, as in my case, I agree that use_assertions should be set globally from a single thread.

**Comment by afabri** [comment]
@GilesBathgate shall I just clean it up?  I ask in order to avoid both/none of us doing it.

**Comment by GilesBathgate** [comment]
> shall I just clean it up?

Please do.

**Comment by GilesBathgate** [comment]
> The only use of that variable right now is to temporarily de-activate Polygon_2 preconditions while checking the validity of a straight skeleton.

@MaelRL The scope of the variable was extended in https://github.com/CGAL/cgal/commit/9b357c6afff4cb7e6175e9096ea49e8585dc9681

**Comment by MaelRL** [comment]
It's a direct translation of https://github.com/CGAL/cgal/blob/efcbc672c015d46929819b77c2b651a97ec67465/Polygon/include/CGAL/Polygon_2/polygon_assertions.h and is still only used in the straight skeleton package. Its purpose is to locally avoid some assertions within a sequential series of functions, which is why it was thread local. Changing it to global can create issues.

**Comment by GilesBathgate** [comment]
I have another idea. For the error handling function, there is `get_static_error_handler()/set_error_handler()`

Maybe there could be something similar for get_use_assertions:

```c++

inline bool& _thread_local_get_use_assertions()
{
  CGAL_STATIC_THREAD_LOCAL_VARIABLE(bool, b, true);
   return b;
}

inline bool& get_use_assertions()
{
  get_use_assertions_handler()();
}

inline Use_assertions_function& get_use_assertions_handler()
{
  static Use_assertions_function _handler = _thread_local_get_use_assertions;
  return _handler;
}

inline void set_use_assertions_handler(Use_assertions_function& f)
{
  get_use_assertions_handler() = f;
}
```

The from user code I can do:

```c++
static bool& useAssertions(){
    /...
}

CGAL::set_use_assertions_handler(useAssertionsHandler);
```

Which also gives the flexibility to use whatever mechanism they want for storing the state.

**Comment by GilesBathgate** [comment]
Yet another option:

```c++

#if defined(CGAL_ENABLE_DISABLE_ASSERTIONS_AT_RUNTIME)

//...

#elif !defined(USER_DEFINED_USE_ASSERTIONS)

namespace CGAL{
inline void set_use_assertions(bool){}
inline constexpr bool get_use_assertions(){return true;}
}

#endif

```

**Comment by afabri** [comment]
In the [code](https://github.com/CGAL/cgal/blob/master/STL_Extension/include/CGAL/assertions.h#L47)  in master we have what @GilesBathgate suggested above.   @lrineau and @sloriot  is that ok for you?

**Comment by afabri** [comment]
@sloriot told me that he is fine with what @GilesBathgate suggested and what is implemented in #7440

**Comment by afabri** [comment]
So we should remove "Under discussion" @lrineau
